bug: implicit enum selector not resolved in map key bracket access

## Description

Implicit enum selectors (e.g., \`.RED\`) work in most contexts where the enum type is known — assignments, comparisons, function arguments, array literals, when cases, and map value assignments. However, they fail with E3110 when used as a **map key** in bracket syntax, even though the map's key type provides the necessary type context.

## Reproduction

```grayscale
const Color enum {
    RED
    GREEN
    BLUE
}

do main() {
    // Map value with implicit selector — works
    mut names [string:Color] = {:}
    names["primary"] = .RED
    println(names["primary"])

    // Map key with implicit selector — fails
    mut palette [Color:string] = {:}
    palette[.RED] = "#FF0000"
    println(palette[.RED])
}
```

## Expected

Compiles successfully. The compiler knows \`palette\` has key type \`Color\`, so \`.RED\` should resolve to \`Color.RED\`.

## Actual

```
error[E3110]: implicit enum selector '.RED' requires type context; use the full form 'EnumName.RED'
```

## Contexts where implicit selectors work

| Context | Works? |
|---|---|
| Variable assignment (\`mut c Color = .RED\`) | Yes |
| Reassignment (\`c = .GREEN\`) | Yes |
| Comparison (\`c == .GREEN\`) | Yes |
| Function argument (\`takes_color(.BLUE)\`) | Yes |
| Array literal (\`[Color] = {.RED, .GREEN}\`) | Yes |
| When case (\`is .GREEN\`) | Yes |
| Map value (\`names["x"] = .RED\`) | Yes |
| **Map key** (\`palette[.RED]\`) | **No** |

## Workaround

Use the fully qualified form: \`palette[Color.RED]\`.

--- a/grayc/src/typechecker/typechecker.c
+++ b/grayc/src/typechecker/typechecker.c
@@ -6603,7 +6603,15 @@ static GrayType *resolve_expression(TypeChecker *checker, AstNode *node) {
 
     case NODE_INDEX_EXPR: {
         GrayType *left = resolve_expression(checker, node->data.index_expr.left);
+        /* Propagate map key type as expected_type so .VARIANT resolves */
+        GrayType *saved_idx_expected = checker->expected_type;
+        if (left->kind == TK_MAP && left->key_type) {
+            GrayType *key_t = typechecker_type_from_name(checker, left->key_type);
+            if (key_t && key_t->kind == TK_ENUM && key_t->name)
+                checker->expected_type = key_t;
+        }
         GrayType *idx_t = resolve_expression(checker, node->data.index_expr.index);
+        checker->expected_type = saved_idx_expected;
         /* E3003: array index must be integer */
         if (left->kind == TK_ARRAY && idx_t->kind != TK_UNKNOWN &&
             !is_int_kind(idx_t->kind) && idx_t->kind != TK_BYTE) {

--- a/integration-tests/pass/core/implicit_enum_selector.gray
+++ b/integration-tests/pass/core/implicit_enum_selector.gray
@@ -170,6 +170,27 @@ do main() {
         failed += 1
     }
 
+    // Test 12: Map key bracket access (assignment)
+    mut palette [Color:string] = {:}
+    palette[.RED] = "#FF0000"
+    if palette[Color.RED] == "#FF0000" {
+        println("  [PASS] map key bracket assignment")
+        passed += 1
+    } otherwise {
+        println("  [FAIL] map key bracket assignment")
+        failed += 1
+    }
+
+    // Test 13: Map key bracket access (read)
+    mut label string = palette[.RED]
+    if label == "#FF0000" {
+        println("  [PASS] map key bracket read")
+        passed += 1
+    } otherwise {
+        println("  [FAIL] map key bracket read")
+        failed += 1
+    }
+
     // Summary
     println("")
     println("Results: ${passed} passed, ${failed} failed")

